aten/src/ATen/native/transformers/hip/flash_attn/flash_api.hip

```diff
@@ -0,0 +1,651 @@
+    else if (d_head == 128)                                                                                                 \
+      CALL_BWD_DROPOUT(FP, BLOCK_M, 128, BLOCK_N, CAUSAL);                                                                  \
+  } while (0)
+
+  if (q_dtype == at::kHalf) {
+    if (is_causal) {
+      CALL_BWD_DROPOUT_DMODEL(__fp16, 16, 16, true);
+    } else {
+      CALL_BWD_DROPOUT_DMODEL(__fp16, 16, 16, false);
+    }
+  } else if (q_dtype == at::kBFloat16) {
+    if (is_causal) {
+      CALL_BWD_DROPOUT_DMODEL(__bf16, 16, 16, true);
+    } else {
+      CALL_BWD_DROPOUT_DMODEL(__bf16, 16, 16, false);
+    }
+  }
+
+  //undo reorder tensors for returns
+  dq = dq_t.permute({0,2,1,3}).contiguous();
+  dk = dk_t.permute({0,2,1,3}).contiguous();
+  dv = dv_t.permute({0,2,1,3}).contiguous();
+
+  // For MQA/GQA we need to sum dK and dV across the groups
+  if (num_heads_k != num_heads) {
+    at::sum_out(dk, at::reshape(dk_expanded, {batch_size, seqlen_k, num_heads_k, num_heads / num_heads_k, head_size}), {3});
+    at::sum_out(dv, at::reshape(dv_expanded, {batch_size, seqlen_k, num_heads_k, num_heads / num_heads_k, head_size}), {3});
+  }
+  return { dq, dk, dv, softmax_d };
+#undef CALL_BWD_DROPOUT
+#undef CALL_BWD
+}
+
+std::tuple<at::Tensor, at::Tensor, at::Tensor, at::Tensor>
+mha_varlen_bwd(const at::Tensor &dout,  // total_q x num_heads, x head_size
+               const at::Tensor &q,   // total_q x num_heads x head_size, total_q := \sum_{i=0}^{b} s_i
+               const at::Tensor &k,   // total_k x num_heads_k x head_size, total_k := \sum_{i=0}^{b} s_i
+               const at::Tensor &v,   // total_k x num_heads_k x head_size, total_k := \sum_{i=0}^{b} s_i
+               const at::Tensor &out,   // total_q x num_heads x head_size
+               const at::Tensor &softmax_lse,     // b x h x s   softmax logsumexp
+               c10::optional<at::Tensor> &dq_,   // total_q x num_heads x head_size, total_q := \sum_{i=0}^{b} s_i
+               c10::optional<at::Tensor> &dk_,   // total_k x num_heads_k x head_size, total_k := \sum_{i=0}^{b} s_i
+               c10::optional<at::Tensor> &dv_,   // total_k x num_heads_k x head_size, total_k := \sum_{i=0}^{b} s_i
+               const at::Tensor &cu_seqlens_q,  // b+1
+               const at::Tensor &cu_seqlens_k,  // b+1
+               const int max_seqlen_q,
+               const int max_seqlen_k,          // max sequence length to choose the kernel
+               const float p_dropout,         // probability to drop
+               const float softmax_scale,
+               const bool zero_tensors,
+               const bool is_causal,
+               const int window_size_left,
+               int window_size_right,
+               const at::Tensor philox_seed,
+               const at::Tensor philox_offset) {
+  TORCH_WARN("[mha_varlen_bwd not supported on ROCm");
```

Comment:
I think this should be an error not a warn no?

Reply:
Replaced with `TORCH_CHECK(false, )` in the latest commit.